src/nvim/api/vim.c

```diff
@@ -1245,6 +1245,225 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
   return keymap_array(mode, NULL);
 }
 
+/// Sets a |mapping| for the given mode.
+///
+/// Note that isn't possible to specify a target buffer when {opts} contains
+/// the key `"buffer"`; like with |:map-<buffer>|, only the current buffer will
+/// be affected.
+///
+/// @param  mode  Mode short-name (the first character of an map command,
+///               e.g. "n", "i", "v", "x", etc.) OR the string "!" (for
+///               |:map!|). |:map| can be represented with an empty string, a
+///               single space " ", or "m".
+/// @param  maptype   Whether to |:map|, |:unmap|, or |:noremap|, represented by
+///                   an empty string, "u", and "n", respectively.
+/// @param  lhs   Left-hand-side |{lhs}| of the mapping.
+/// @param  rhs   Right-hand-side |{rhs}| of the mapping.
+/// @param  opts  Optional parameters. Includes all |:map-arguments| as keys.
+///               Values should all be Booleans. Unrecognized keys will result
+///               in an error.
+/// @param[out]   err   Error details, if any.
+///
+/// @returns  Zero on success, nonzero on failure.
+Integer nvim_set_keymap(String mode, String maptype, String lhs, String rhs,
+                        Dictionary opts, Error *err)
+  FUNC_API_SINCE(6)
+{
+  // maximum possible size of the args-string we pass to do_map()
+  // allocate memory right away, before any calls to free
+  static const size_t kMaxSpecialArgSize =
+      sizeof("<buffer> <nowait> <silent> <script> <expr> <unique>");
+  // leave room for spaces around lhs, rhs; and null character
+  const size_t kMaxArgSize = kMaxSpecialArgSize + lhs.size + rhs.size + 3;
+  char_u *args = (char_u *)xcalloc(kMaxArgSize, sizeof(char_u));
+
+  char *err_msg = NULL;  // the error message to report, if any
+  char *err_arg = NULL;  // argument for the error message format string
+  ErrorType err_type = kErrorTypeNone;
+
+  // make sure that lhs and rhs aren't purely whitespace, and don't contain
```

Comment:
`rhs` and `lhs` should ideally allow whitespace, this is really just a parsing issue of `do_map`.  A strategy could be to split `do_map` after the parsing, probaly after the `if (maptype == 1 !haskey) {` condition. Then this API function could skip the whitespace parsing entirely, and send in lhs/rhs with whitespace in it (which then means the same as `<space>`) . Though maybe `\n` should still be forbidden, it litteraly is the same as `<c-j>` but many users would probably expect it to be mean `<cr>`